flight/Modules/VtolPathFollower/vtol_follower_control.c

```diff
@@ -72,123 +197,184 @@ int32_t vtol_follower_control_path(const float dT, const PathDesiredData *pathDe
 {
 	PositionActualData positionActual;
 	PositionActualGet(&positionActual);
+
+	VelocityActualData velocityActual;
+	VelocityActualGet(&velocityActual);
+
+	const float cur_pos_ned[3] = {
+		positionActual.North +
+		    velocityActual.North*guidanceSettings.PositionFeedforward,
+		positionActual.East +
+		    velocityActual.East*guidanceSettings.PositionFeedforward,
+		positionActual.Down };
+
+	path_progress(pathDesired, cur_pos_ned, progress);
 	
-	const float cur[3] = { positionActual.North, positionActual.East, positionActual.Down };
-	
-	path_progress(pathDesired, cur, progress);
-	
+	// Interpolate desired velocity and altitude along the path
+	float groundspeed = vtol_interpolate(progress->fractional_progress,
+	    pathDesired->StartingVelocity, pathDesired->EndingVelocity);
+
+	if (progress->fractional_progress >= 1)
+		groundspeed = 0;
+
+	float altitudeSetpoint = vtol_interpolate(progress->fractional_progress,
+	    pathDesired->Start[2], pathDesired->End[2]);
+
+	float commands_ned[3];
+
+	float downError = altitudeSetpoint - positionActual.Down;
+	commands_ned[2] = pid_apply_antiwindup(&vtol_pids[DOWN_POSITION], downError,
+		-guidanceSettings.VerticalVelMax, guidanceSettings.VerticalVelMax, dT);
+
 	// Update the path status UAVO
 	PathStatusData pathStatus;
 	PathStatusGet(&pathStatus);
 	pathStatus.fractional_progress = progress->fractional_progress;
 	pathStatus.error = progress->error;
-	if (pathStatus.fractional_progress < 1)
+	pathStatus.Waypoint = pathDesired->Waypoint;
+
+	// Only if we are at the end, and above the end altitude, are we
+	// done.
+	if ((pathStatus.fractional_progress < 1) ||
+		(downError < -guidanceSettings.EndpointRadius)) {
 		pathStatus.Status = PATHSTATUS_STATUS_INPROGRESS;
-	else
+	} else {
 		pathStatus.Status = PATHSTATUS_STATUS_COMPLETED;
+	}
 	PathStatusSet(&pathStatus);
 
-	float groundspeed = pathDesired->StartingVelocity + 
-	    (pathDesired->EndingVelocity - pathDesired->StartingVelocity) * progress->fractional_progress;
-	if (progress->fractional_progress > 1)
-		groundspeed = 0;
-	
-	VelocityDesiredData velocityDesired;
-	velocityDesired.North = progress->path_direction[0] * groundspeed;
-	velocityDesired.East = progress->path_direction[1] * groundspeed;
-	
-	float error_speed = progress->error * 
+	if (groundspeed == 0) {
+		return vtol_follower_control_simple(dT, pathDesired->End, 0, 1);
```

Comment:
this is double settings the status is it not. also use boolean true/false instead of 1/0